app/controllers/application_controller.rb

```diff
@@ -11,4 +11,22 @@ def after_sign_in_path_for(user)
     pages_url
   end
 
+  private
+
+  def localize_from_page_id
+    page = Page.find_by(id: params[:page_id])
+    if page.present? && page.language.present? && page.language.code.present?
+      set_locale(page.language.code)
+    end
+  end
+
+  def set_locale(code)
+    begin
+      I18n.locale = code
+    rescue I18n::InvalidLocale
+      # by setting the +i18n.enforce_available_locales+ flag to true but
+      # catching the resulting error, it allows us to only set the locale
+      # if it's one explicitly registered under +i18n.available_locales+
+    end
```

Comment:
Yes that's pretty much it. But why recreate the `I18n` module's internal logic when we can set it to throw an exception we know we can catch?